sound/soc/codecs/rt715-sdca.c

```diff
@@ -263,30 +273,82 @@ static int rt715_sdca_put_volsw(struct snd_kcontrol *kcontrol,
 	struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
 	struct soc_mixer_control *mc =
 		(struct soc_mixer_control *)kcontrol->private_value;
-	unsigned int val, val2, loop_cnt = 2, i;
-	unsigned int invert = mc->invert;
-	unsigned int reg2 = mc->rreg;
 	unsigned int reg = mc->reg;
-	unsigned int max = mc->max;
+	unsigned int reg2 = mc->rreg;
+	unsigned int shift = mc->shift;
+	unsigned int rshift = mc->rshift;
+	int max = mc->max;
+	int min = mc->min;
+	unsigned int sign_bit = mc->sign_bit;
+	unsigned int mask = (1 << fls(max)) - 1;
+	unsigned int invert = mc->invert;
 	int err;
+	bool type_2r = false;
+	unsigned int val[4] = {0}, val_mask;
+	unsigned int loop_cnt = 2, i;
 
-	val = ucontrol->value.integer.value[0];
-	if (invert)
-		val = max - val;
-
-	val2 = ucontrol->value.integer.value[1];
-	if (invert)
-		val2 = max - val2;
+	if (sign_bit)
+		mask = BIT(sign_bit + 1) - 1;
 
 	for (i = 0; i < loop_cnt; i++) {
-		err = snd_soc_component_write(component, reg + i * 2, val);
-		if (err < 0)
-			return err;
-		err = snd_soc_component_write(component, reg2 + i * 2, val2);
+		val[i * 2] = ((ucontrol->value.integer.value[i * 2] + min) & mask);
+		if (invert)
+			val[i * 2] = max - val[i * 2];
+		val_mask = mask << shift;
+		val[i * 2] = val[i * 2] << shift;
+
+		val[i * 2 + 1] =
+			((ucontrol->value.integer.value[i * 2 + 1] + min) & mask);
+		if (invert)
+			val[i * 2 + 1] = max - val[i * 2 + 1];
+		if (reg == reg2) {
+			val_mask |= mask << rshift;
+			val[i * 2] |= val[i * 2 + 1] << rshift;
+		} else {
+			val[i * 2 + 1] = val[i * 2 + 1] << shift;
+			type_2r = true;
+		}
+
+		err = snd_soc_component_update_bits(component, reg + i * 2, val_mask,
+				val[i * 2]);
 		if (err < 0)
 			return err;
+
+		if (type_2r)
+			err = snd_soc_component_update_bits(component, reg2 + i * 2,
+				val_mask, val[i * 2 + 1]);
 	}
 
+	return err;
+}
+
+static int rt715_sdca_fu_info(struct snd_kcontrol *kcontrol,
+	struct snd_ctl_elem_info *uinfo)
+{
+	uinfo->type = SNDRV_CTL_ELEM_TYPE_BOOLEAN;
+	uinfo->count = 4;
+	uinfo->value.integer.min = 0;
+	uinfo->value.integer.max = 1;
+	return 0;
+}
+
+static int rt715_sdca_vol_info(struct snd_kcontrol *kcontrol,
+	struct snd_ctl_elem_info *uinfo)
+{
+	uinfo->type = SNDRV_CTL_ELEM_TYPE_INTEGER;
+	uinfo->count = 4;
+	uinfo->value.integer.min = 0;
+	uinfo->value.integer.max = 0x7f;
```

Comment:
This does not look correct. The max value should be 1,2 or 4 like in rt715_sdca_set_amp_gain_get(). It would make sense to put the common code (obtain channel count) to a separate function.

Reply:
Oops. I meant the _uinfo->count_.

Reply:
ok, I'll put the common code to a separate function according to its max value.

Reply:
The updated version https://github.com/thesofproject/linux/commit/83fa41c5b2b760d613c2056ad436f64c598a9fc3 is even worse - you duplicated big code chunks. I meant to create a new function for this code:

<pre>
if (strstr(ucontrol->id.name, "FU02 Capture Volume") ||
		strstr(ucontrol->id.name, "FU06 Capture Volume"))
		loop_cnt = 2;
	else if (strstr(ucontrol->id.name, "FU0E Boost") ||
		strstr(ucontrol->id.name, "FU0C Boost"))
		loop_cnt = 4;
	else
		loop_cnt = 1;
</pre>

Reply:
@jack-cy-yu Here is an example

```
struct rt715_sdca_kcontrol_private {
       unsigned int reg_base;
       unsigned int count;
       unsigned int max;
};

#define RT715_SDCA_BOOST_EXT_TLV_VALUE(xreg_base, xcount, xmax) \
       ((unsigned long)&(struct rt715_sdca_kcontrol_private) \
        {.reg_base = xreg_base, .count = xcount, .max = xmax})

#define RT715_SDCA_BOOST_EXT_TLV(xname, reg_base, xhandler_get,\
         xhandler_put, tlv_array, xcount, xmax) \
{       .iface = SNDRV_CTL_ELEM_IFACE_MIXER, .name = (xname), \
        .access = SNDRV_CTL_ELEM_ACCESS_TLV_READ | \
                 SNDRV_CTL_ELEM_ACCESS_READWRITE, \
        .tlv.p = (tlv_array), \
        .count = xcount, \
        .info = rt715_sdca_boost_info, \
        .get = xhandler_get, .put = xhandler_put, \
        .private_value = RT715_SDCA_BOOST_EXT_TLV_VALUE(reg_base, xcount, xmax) }

static int rt715_sdca_boost_info(struct snd_kcontrol *kcontrol,
        struct snd_ctl_elem_info *uinfo)
{
        struct rt715_sdca_kcontrol_private *p =
                (struct rt715_sdca_kcontrol_private *)kcontrol->private_value;

        uinfo->type = SNDRV_CTL_ELEM_TYPE_INTEGER;
        uinfo->count = p->count;
        uinfo->value.integer.min = 0;
        uinfo->value.integer.max = p->max;
        return 0;
}

static int rt715_sdca_set_amp_gain_8ch_get(struct snd_kcontrol *kcontrol,
                struct snd_ctl_elem_value *ucontrol)
{
        struct snd_soc_component *component = snd_kcontrol_chip(kcontrol);
        struct rt715_sdca_priv *rt715 = snd_soc_component_get_drvdata(component);
        struct rt715_sdca_kcontrol_private *p =
                (struct rt715_sdca_kcontrol_private *)kcontrol->private_value;
        unsigned int reg_base = p->reg_base;
}
```
Hope it helps.